flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/iomanager/AsynchronousBufferFileWriter.java

```diff
@@ -31,9 +31,26 @@ protected AsynchronousBufferFileWriter(ID channelID, RequestQueue<WriteRequest>
 		super(channelID, requestQueue, CALLBACK, true);
 	}
 
+	/**
+	 * Writes the given block asynchronously.
+	 *
+	 * @param buffer
+	 * 		the buffer to be written (will be recycled when done)
```

Comment:
Shouldn't this contract be in `BlockChannelWriterWithCallback`? Now it's kind of strange that recycle is implementation dependent. Especially that `SpillableSubpartition` is using `BufferFileWriter` and still depends on recycling.

Reply:
good catch, but actually, `SpillableSubpartition` doesn't do any recycling itself: in its `finish()` method, it relies on the buffer being on-heap and then garbage-collected, for the `add()` function, it relies on the caller, i.e. `ResultPartition#add()` (which I also forgot to adapt).